README.md
# Terminal Markdown Viewer

[![Build Status][travis_img]][travis]
<a href='https://coveralls.io/github/axiros/terminal_markdown_viewer?branch=master'>
<img src='https://coveralls.io/repos/github/axiros/terminal_markdown_viewer/badge.svg?branch=master' alt='Coverage Status' /></a>
[![PyPI version](https://badge.fury.io/py/mdv.svg)](https://badge.fury.io/py/mdv)
<a href="https://github.com/ambv/black"><img alt="Code style: black" src="https://img.shields.io/badge/code%20style-black-000000.svg"></a>




When you edit multiple md files remotely, like in a larger
[mkdocs](http://www.mkdocs.org/) project, context switches between editing
terminal(s) and viewing browser may have some efficiency impact.
Also sometimes there is just no browser, like via security gateways offering
just a fixed set of applications on the hop in machine.
Further, reading efficiency and convenience is often significantly improved
by using colors.
And lastly, using such a thing for cli applications might improve user output,
e.g. for help texts.

This is where mdv, a Python based Markdown viewer for the terminal might be
a good option.

<!-- toc -->

- [Terminal Markdown Viewer](#terminal-markdown-viewer)
	- [Features](#features)
	- [Alternatives](#alternatives)
	- [Installation](#installation)
	- [Usage](#usage)
		- [CLI](#cli)
		- [Inline](#inline)
		- [Sample Inline Use Case: click module docu](#sample-inline-use-case-click-module-docu)
	- [Customization](#customization)
	- [Screenshots](#screenshots)
	- [TODO](#todo)
	- [Credits](#credits)
	- [Updates](#updates)
		

<!-- tocstop -->



If markdown is often "simple" enough to be somewhat readable on 256 color terminals (except images that is).

<img src="./samples/1.png" width=500>

from

	### Source
	# Header 1
	## Header 2
	### Header 3
	#### Header 4
	##### Header 5
	###### Header 6
	```python
	""" Test """
	# Make Py2 >>> Py3:
	import os, sys; reload(sys); sys.setdefaultencoding('utf-8')
	# no? see http://stackoverflow.com/a/29832646/4583360 ...

	# code analysis for hilite:
	try:
	    from pygments import lex, token
	    from pygments.lexers import get_lexer_by_name, guess_lexer
	```

	| Tables | Fmt |
	| -- | -- |
	| !!! hint: wrapped | 0.1 **strong** |

	!!! note: title
	    this is a Note


You can also use mdv as a **source code** viewer, best when you have docstrings with markdown in your code:

![](./samples/5.png)

from

```python
~/terminal_markdown_viewer $ cat setup.py
#!/usr/bin/env python2.7
# coding: utf-8

"""_
# Mdv installation

## Usage

    [sudo] ./setup.py install

----
"""

from setuptools import setup, find_packages

import mdv

setup(
    name='mdv',
    version=mdv.__version__,

```
(the '_' after the docstring telling mdv that markdown follows)

----

> mdv is a proof of concept hack: While for simple structures it does its job quite well, for complex markdown you want to use other tools.
> Especially for inlined html it simply fails.

----


## Features

- Tons of theme combinations: mdv ships with > 200 luminocity sorted themes, converted from html themes tables to ansi. Those can be combined for code vs regular markdown output...
- Admonitions
- Tables, incl. wide table handling avoiding "interleaving"
- Somewhat hackable, all in [one](mdv/markdownviewer.py) module
- Useable as lib as well
- File change monitor
- Text wrapping
- Source code highlighter
- Little directory change monitor (cames handy when working on multiple files, to get the current one always displayed)
	- which can run arbitrary commands on file changes
	- which passes filepath, raw and prettyfied content to the other command
        Note: Poor man's implementation, polling. Check inotify based tools if you want sth better.

## Alternatives

The ones I know of (and which made me write mdv ;-) ):

1. There are quite a few from the js community (e.g. [msee](https://www.npmjs.com/package/msee), ansidown, ansimd and also nd which is great) but they require nodejs & npm, which I don't have on my servers. Also I personally wanted table handling and admonition support throughout and prob. too old to hack other peoples' js (struggling enough with my own). But have a look at them, they do some things better than mdv in this early version (I try to learn from them). Also [this](https://github.com/substack/picture-tube) would be worth a look ;-)
2. pandoc -> html -> elinks, lynx or pandoc -> groff -> man. (Heavy and hard to use from within other programs. Styling suboptimal)
3. vimcat (Also heavy and hard to use inline in other programs)

Summary: For production ready robust markdown viewing (e.g. for your customers) I recommend nd still, due to the early state of mdv. For playing around, especially with theming or when with Python, this one might be a valid alternative to look at.

## Installation

    pip install mdv

If you get `no attribute HTML_PLACEHOLDER`: update your markdown package.

[Here](https://trac.macports.org/ticket/53591) is a macport (thanks Aljaž).

mdv is also available in the FreeBSD package repositories via `pkg install py36-mdv` (the Python version might change in the future).

### Manual Install: Requirements

- python == 2.7 or > 3.5
- py markdown (pip install markdown)
- py pygments (pip install pygments)
- py yaml (pip install pyyaml)
- py docopt (pip install docopt)
- py tabulate (pip install tabulate)

Further a 256 color terminal (for now best with dark background) and font support for a few special separator characters (which you could change via config).

> For light terms you'd just need to revert the 5 colors from the themes, since they are sorted by luminocity.

I did not test anything on windows.

## Usage

### CLI

```markdown

# Usage:

    mdv [OPTIONS] MDFILE

# Options:

    MDFILE    : Path to markdown file
    -A        : Strip all ansi (no colors then)
    -C MODE   : Sourcecode highlighting mode
    -H        : Print html version
    -L        : Backwards compatible shortcut for '-u i'
    -M DIR    : Monitor directory for markdown file changes
    -T C_THEME: Theme for code highlight. If not set: Using THEME.
    -X Lexer  : Default lexer name (default: python). Set -x to use it always.
    -b TABL   : Set tab_length to sth. different than 4 [default: 4]
    -c COLS   : Fix columns to this (default: your terminal width)
    -f FROM   : Display FROM given substring of the file.
    -h        : Show help
    -i        : Show theme infos with output
    -l        : Light background (not yet supported)
    -m        : Monitor file for changes and redisplay FROM given substring
    -n NRS    : Header numbering (default: off. Say e.g. -3 or 1- or 1-5
    -t THEME  : Key within the color ansi_table.json. 'random' accepted.
    -u STYL   : Link Style (it=inline table=default, h=hide, i=inline)
    -x        : Do not try guess code lexer (guessing is a bit slow)


# Notes:

We use stty tool to derive terminal size. If you pipe into mdv we use 80 cols.

## To use mdv.py as lib:

Call the main function with markdown string at hand to get a
formatted one back. Sorry then for no Py3 support, accepting PRs if they don't screw Py2.

## FROM:

FROM may contain max lines to display, seperated by colon.
Example:

    -f 'Some Head:10' -> displays 10 lines after 'Some Head'

If the substring is not found we set it to the *first* character of the file -
resulting in output from the top (if your terminal height can be derived correctly through the stty cmd).

## Code Highlighting

Set -C <all|code|doc|mod> for source code highlighting of source code files.
Mark inline markdown with a '_' following the docstring beginnings.

- all: Show markdown docstrings AND code (default if you say, e.g. `-C.`)
- code: Only Code
- doc: Only docstrings with markdown
- mod: Only the module level docstring


## File Monitor:

If FROM is not found we display the whole file.

## Directory Monitor:

We check only text file changes, monitoring their size.

By default .md, .mdown, .markdown files are checked but you can change like `-M 'mydir:py,c,md,'` where the last empty substrings makes mdv also monitor any file w/o extension (like 'README').

### Running actions on changes:

If you append to `-M` a `'::<cmd>'` we run the command on any change detected (sync, in foreground).

The command can contain placeholders:

    _fp_     # Will be replaced with filepath
    _raw_    # Will be replaced with the base64 encoded raw content
               of the file
    _pretty_ # Will be replaced with the base64 encoded prettyfied output

Like: mdv -M './mydocs:py,md::open "_fp_"'  which calls the open
command with argument the path to the changed file.


## Themes

### Theme Rollers


    mdv -T all [file]:  All available code styles on the given file.
    mdv -t all [file]:  All available md   styles on the given file.
                        If file is not given we use a short sample file.

So to see all code hilite variations with a given theme:

Say C_THEME = all and fix THEME

Setting both to all will probably spin your beach ball...

### Environ Vars

`$MDV_THEME` and `$MDV_CODE_THEME` are understood, e.g. `export
MDV_THEME=729.8953` in your .bashrc will give you a consistent color scheme.


```

> Regarding the strange theme ids: Those numbers are the calculated total luminocity of the 5 theme colors.

### Inline

mdv is designed to be used well from other (Py2) programs when they have md at hand which should be displayed to the user:

```python
import mdv

# config like this:
mdv.term_columns = 60

# calling like this (all CLI options supported, check def main
formatted = mdv.main(my_raw_markdown, c_theme=...)  
```

> Note that I set the defaultencoding to utf-8  in ``__main__``. I have this as my default python2 setup and did not test inline usage w/o. Check [this](http://stackoverflow.com/a/29832646/4583360) for risks.

### Sample Inline Use Case: click module docu

[Armin Ronacher](http://lucumr.pocoo.org/2014/5/12/everything-about-unicode/)'s
[click](http://click.pocoo.org) is a great framework for writing larger CLI apps - but its help texts are a bit boring, intended to be customized.

Here is how:

Write a normal click module with a function but w/o a doc string as shown:
```python
@pass_context                                                                   
def cli(ctx, action, name, host, port, user, msg):           
	""" docu from module __doc__ """
```

On module level you provide markdown for it, like:

```shell
~/axc/plugins/zodb_sub $ cat zodb.py | head
"""
# Fetch and push ZODB trees

## ACTION: < info | pull | push | merge | dump | serve>

- info:  Requests server availability information
(...)
```
which you set at click module import time:

	mod.cli.help = mod.__doc__


Lastly do this in your app module:

```python
from click.formatting import HelpFormatter
def write_text(self, text):
    """ since for markdown pretty out on cli I found no good tool
	so I built my own """
    # poor man's md detection:
    if not text.strip().startswith('#'):
        return orig_write_text(self, text)
    from axc.markdown.mdv import main as mdv
    self.buffer.append(mdv(md=text, theme=os.environ['AXC_THEME']))

HelpFormatter.orig_write_text = HelpFormatter.write_text
HelpFormatter.write_text = write_text
```

The output has then colors:

![](samples/3.png)

and at smaller terms rewraps nicely:

![](samples/4.png)

Further, having markdown in the module ``__doc__`` makes it simple to add into a global project docu framework, like mkdocs.



## Customization

You can supply all CLI args in `$HOME/.mdv`, in yaml format.

More flex you have via `$HOME/.mdv.py`, which is execed if present, when
running `main`.

Alternatively, in [mdv.py](mdv.py) you can change some config straight forward.

```python
# ---------------------------------------------------------------------- Config
txt_block_cut, code_pref, list_pref, br_ends = '✂', '| ', '- ', '◈'
# ansi cols (default):
# R: Red (warnings), L: low visi, BG: background, BGL: background light, C=code
# H1 - H5 = the theme, the numbers are the ansi color codes:
H1,  H2,  H3,  H4,  H5, R,   L,  BG, BGL, T,   TL, C   = \
231, 153, 117, 109, 65, 124, 59, 16, 188, 188, 59, 102
# Code (C is fallback if we have no lexer). Default: Same theme:
CH1, CH2, CH3, CH4, CH5 = H1, H2, H3, H4, H5

code_hl = { "Keyword" : 'CH3', "Name" : 'CH1',
            "Comment" : 'L',  "String": 'CH4',
            "Error"   : 'R',  "Number": 'CH4',
            "Operator": 'CH5',
            "Generic" : 'CH2'
            }

admons = {'note'     : 'H3', 'warning': 'R',
          'attention': 'H1', 'hint'   : 'H4',
          'summary'  : 'H1', 'hint'   : 'H4',
          'question' : 'H5', 'danger' : 'R',
          'caution'  : 'H2'
         }

def_lexer = 'python'
guess_lexer = True
# also global. but not in use, BG handling can get pretty involved...
background = BG

# normal text color:
color = T

show_links = None

# could be given, otherwise read from ansi_tables.json:
themes = {}


# sample for the theme roller feature:
md_sample = ''

# ------------------------------------------------------------------ End Config
```

Any importing module can overwrite those module global variables as well.

Should you need yet additional themes, add them to ``ansi_tables.json`` file by adding your ansi codes there.



## Screenshots

Random results, using the theme roller feature:

![second](https://github.com/axiros/terminal_markdown_viewer/blob/master/samples/2.png)

Note the table block splitting when the table does not fit (last picture).

## TODO

- Refactor the implementation, using a config class
- Lines separators not optimal ([nd](https://www.npmjs.com/package/nd) does better)
- Test light colorscheme
- Dimming
- A few grey scale and 8 color themes
- Sorting of the json by luminance
- Some themes have black as darkest color, change to dark grey
- Common Mark instead of markdown

## PerfTests

Rendering this readme [100 times](./mdv/misc/perftest.py):
```
black root@ip-10-34-2-19:~/terminal_markdown_viewer/mdv/misc# python perfest.py
0.03 paka
0.04 paka_breaks
0.04 paka_xml
1.47 mistletoe
8.70 markdown
5.22 commonmark
```
- markdown did better than commonmark w/o extensions but table and fenced code
are definitelly required for 99% users.

- paka is a wrapper around the C reference lib -> requires compilation.

- mistletoe is pure python, crazy that they are so much faster than CommonMark.
They say in pypy they are speed up even much more.

mistletoe downside: py2 only via a fork.


## Credits

[pygments](http://pygments.org/) (using their lexer)

[tabulate](https://pypi.python.org/pypi/tabulate)

and, naturally, the [python markdown project](https://pythonhosted.org/Markdown/authors.html)

Update: Next version will be CommonMark based though...


## Updates

### July 2016:

Sort of an excuse for the long long time w/o an update:
I did actually start working on a more solid version based on CommonMark but
that went a bit out of scope, into a general html terminal viewer, which will
probably never be finished :-/

So at least here an update containing the stuff you guys sent as PRs, thanks all!!

- installation and dependencies via a setup.py (thanks
  [Martin](https://github.com/althonos))
- supporting `echo -e "# foo\n## bar" | mdv -` and a 'light' theme (thanks
  [Stanislav](https://github.com/seletskiy))
- and a few other improvements regarding python2.7, file location and pyyaml, thanks all.

Also:

- fixed the most obvious bugs with nested ordered and unordered lists
- fixed bold marker
- different color highlighting for the list markers
- added a source code highlighting mode, which highlights also docstrings in markdown (`-C <mode>`)
- some tests in the tests folder
- using `textwrap` now for the wrapping, to avoid these word breaks a few complained about
- you can supply the default lexer now, e.g. `-X javascript [-x]`
- fixed but with not rendered strong texts
- pip install mdv


### Nov 2016:

- travis

- Inline link tables

![](./samples/links.png)



[travis]: https://travis-ci.org/axiros/terminal_markdown_viewer
[travis_img]: https://travis-ci.org/axiros/terminal_markdown_viewer.svg?branch=master



### Sept 2018:

- Merged some PRs, thanks.
- Decent [code formatter](https://github.com/ambv/black). Not that this weekend hack got more readable though. Well, maybe a bit.
- Revised Py3 support (finally found peace with it, since they enforce UTF-8 everywhere the new features begin to outweigh the nightmares of trying to decode everything without need).
- Indented code in PY3 was broken, fixed that. Why, PY3, are you creating crap like `"b'foo'"` instead raising or auto-decoding (since you work anyway only with your UTF8-everywhere-assumption)!?
- Header numbering feature added (`-n 2-4` or `-n 1-`)
<img src="./samples/header_num.png" width="400"/>
- docopt and pyyaml install requirement removed, better config file handling.
- pypi markdown rendering for the readme, finally.
<!-- pre_proc_marker -->


----


Last modified: Fri Jul 30 11:34:23 2021

CHANGELOG.md
<!-- AUTOMATICALLY GENERATED FILE - DO NOT DIRECTLY EDIT!

Direct edits will be gone after next CI build.
By: gk@axgk (Thu Jul 29 20:11:23 2021)
Command Line (see duties.py):

    /home/gk/miniconda3/envs/mdv_py37/bin/doc pre_process \
     --patch_mkdocs_filewatch_ign_lp \
     --gen_theme_link \
     --gen_last_modify_date \
     --gen_change_log \
     --gen_change_log_versioning_stanza=semver \
     --gen_change_log \
     --gen_credits_page \
     --gen_auto_docs \
     --lit_prog_evaluation=md \
     --lit_prog_evaluation_timeout=5 \
     --lit_prog_on_err_keep_running=false
-->

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.7.0](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.7.0)
<small>[Compare with 1.6.3](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.6.3...1.7.0)</small>

### Bug Fixes
- Devplugins match new module naming ([e60d200](https://github.com/AXGKl/terminal_markdown_viewer/commit/e60d200886048f6065878f8bef50384080cdf82f) by Gunther Klessinger).
- Issue/93 monitor_dir ([435b688](https://github.com/AXGKl/terminal_markdown_viewer/commit/435b6885d0237fe10a9f894375b10c1103bfaae0) by Gunther Klessinger).
- Dockerfile mdv on python3 and install via pip install -e ([d268f4f](https://github.com/AXGKl/terminal_markdown_viewer/commit/d268f4fac6837e8e5ce3e2c9aa9185d4d5f7c027) by Gunther Klessinger).
- Some fixes for python3 from prs ([d2e8d26](https://github.com/AXGKl/terminal_markdown_viewer/commit/d2e8d26f39590fdc9c0b9ec0b80b578c7e260c6c) by Gunther Klessinger).
- Add some no cover to pass coverage ([90c3736](https://github.com/AXGKl/terminal_markdown_viewer/commit/90c37360eb5f60166b80baff43c8c9d8a3ff1bb8) by rachmadaniHaryono).
- Dev: setdefaultencoding not on py3. ([eb56c2f](https://github.com/AXGKl/terminal_markdown_viewer/commit/eb56c2f32957c51e7e85ba1024c93301b3f564a3) by rachmadaniHaryono).

### Features
- Passing non config args into run functions, allowing easy params ([b5abd40](https://github.com/AXGKl/terminal_markdown_viewer/commit/b5abd403b7da2c7aee4bb55e2d33a490d054dca1) by Gunther Klessinger).
- Html_out, term_out, ruler, internal style tags ([8b3a676](https://github.com/AXGKl/terminal_markdown_viewer/commit/8b3a6764a44f1ee734f0b4664411ca5ff2b7765d) by Gunther Klessinger).
- Little dev/maintenance helper ([47b1bb8](https://github.com/AXGKl/terminal_markdown_viewer/commit/47b1bb82d92f06b83a2de30233cda3c7773e4cc6) by Gunther Klessinger).
- V2, with css and html support ([931de61](https://github.com/AXGKl/terminal_markdown_viewer/commit/931de61456433d646829ec0f62c33dcdd93b6b20) by Gunther Klessinger).


## [1.6.3](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.6.3)
<small>[Compare with 1.6.2](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.6.2...1.6.3)</small>


## [1.6.2](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.6.2)
<small>[Compare with 1.6.1](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.6.1...1.6.2)</small>


## [1.6.1](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.6.1)
<small>[Compare with 1.6.0](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.6.0...1.6.1)</small>


## [1.6.0](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.6.0)
<small>[Compare with 1.5.0](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.5.0...1.6.0)</small>


## [1.5.0](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.5.0)
<small>[Compare with 1.4.1](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.4.1...1.5.0)</small>


## [1.4.1](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.4.1)
<small>[Compare with 1.4.0](https://github.com/AXGKl/terminal_markdown_viewer/compare/1.4.0...1.4.1)</small>


## [1.4.0](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/1.4.0)
<small>[Compare with 0.4](https://github.com/AXGKl/terminal_markdown_viewer/compare/0.4...1.4.0)</small>


## [0.4](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/0.4)
<small>[Compare with 0.3](https://github.com/AXGKl/terminal_markdown_viewer/compare/0.3...0.4)</small>


## [0.3](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/0.3)
<small>[Compare with v0.2](https://github.com/AXGKl/terminal_markdown_viewer/compare/v0.2...0.3)</small>


## [v0.2](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/v0.2)
<small>[Compare with v0.1.2](https://github.com/AXGKl/terminal_markdown_viewer/compare/v0.1.2...v0.2)</small>


## [v0.1.2](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/v0.1.2)
<small>[Compare with v0.1.1](https://github.com/AXGKl/terminal_markdown_viewer/compare/v0.1.1...v0.1.2)</small>


## [v0.1.1](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/v0.1.1)
<small>[Compare with v0.1](https://github.com/AXGKl/terminal_markdown_viewer/compare/v0.1...v0.1.1)</small>


## [v0.1](https://github.com/AXGKl/terminal_markdown_viewer/releases/tag/v0.1)
<small>[Compare with first commit](https://github.com/AXGKl/terminal_markdown_viewer/compare/fbf8c2b7aa9154841ecc38efde9136b5be35e85d...v0.1)</small>



src/mdv/plugins/config.py
# fmt:off
environ_prefix = 'MDV_'

class Plugins:

    boxes          = 'term_css_boxes'
    color          = 'color'
    colors_256     = 'color_table_256'
    colors_web     = 'color_table_web'
    conf           = 'mdv_conf'
    log            = 'structlog'
    mdparser       = 'pymarkdown'
    render         = 'term_render'
    style          = 'term_css_style'
    textmaps       = 'term_font_textmaps.py'
    theme          = 'theme_base16'
    tree_analyzer  = 'html_beautifulsoup'

    class Actions:
        colortables= 'colortables'
        view       = 'view'
        help       = 'help'

class Output:
    html_out = None # when given we write the html here ('-' for stdout)
    term_out = None # when given we write the ansi here ('-' set as default when run from cli)
    ruler = None # add a horizontal ruler for terminal output
    

class Logging:
    # pip install structlog provides nice dev logging Eats around 0.03s app startup time though.
    log_level     = 'info'

class Terminal:
    true_color     = True
    width          = 0 # 0: auto
    height         = 0
    width_default  = 80 # fallback
    height_default = 24

class Markdown:
    rm_spaces   = True
    tab_length  = 4
    sample_text = '''# H1\nHello *World!* from mdv!'''


class Styling:
    # those are initialized with display: inline even w/o a stylesheet:
    inline_tags = ['em', 'strong', 'b', 'code', 'del', 'super', 'sub']
    theme = 'theme_base16' 
    css_file   = '' # requires pip install cssutils

    class Variables:
        class Text:
            hr_sep        = '─'
            txt_block_cut = '✂'
            code_pref     = '| '
            list_pref     = '- '
            bquote_pref   = '|'
            hr_ends       = '◈'

        class Cells:
            H1     = '1;33'
            H2     = '1;32'
            H3     = '1;35'
            H4     = '1;35'
            H5     = '1;34'
            R      = '31'
            L      = '2;90'
            BG     = '30'
            BGL    = '30'
            D      = ''
            T      = '36'
            C      = '3;100;97'
            EM     = '3;31'
            STRONG = '1;32;41'
            CH1, CH2, CH3, CH4, CH5 = H1, H2, H3, H4, H5

        class Admonitions:
            attention = 'H1'
            caution   = 'H2'
            danger    = 'R'
            dev       = 'H5'
            hint      = 'H4'
            note      = 'H3'
            question  = 'H5'
            summary   = 'H1'
            warning   = 'R'

        class Code:
            def_lexer   = 'python'
            guess_lexer = True

            class Highlight:
                Comment  = 'L'
                Error    = 'R'
                Generic  = 'CH2'
                Keyword  = 'CH3'
                Name     = 'CH1'
                Number   = 'CH4'
                Operator = 'CH5'
                String   = 'CH4'


    class CSS:
        BQ         = {'border': '2 solid H1'}
        UL         = {'padding': 2}
        OL         = {'padding': 2}


# ansi cols (default):
# Convention:
# Numbers are the ansi 256 color codes (those are independent of your terminal theme)
# Strings: Ansi code as is.
# Example: 124 rendered identical as '38;5;124' or 'H4' if H4 then is 124
# The string form allows to set also underlined, bold, (...) if your term supports it.
# See e.g.: https://stackoverflow.com/questions/4842424/list-of-ansi-color-escape-sequences/33206814#33206814

# Semantics:
# R: Red (warnings), L: low visi, BG: background, BGL: background light, C=code
# H1 - H5 = the color theme.


#     ,  # H1: bold yellow
#     ,  # bold green
#     ,  # bold cyan
#     ,  # bold magenta
#     ,  # H5 bold blue
#     ,  # R red
#     ,  # L dimmed bright black -> rendered as gray usually
#     ,  # black
#     ,  # black
#     '',  # D: document (outer div)
#     ,  # T(ext) white
#     ,  # C(ode) italic bright b/w bg and fg
#     '3;31',  # em italics
#     '1;32;41',  # bold
#     {'border': '2 solid H1'},
#     # {
#     #     'ansi': '1;32',
#     #     'border': '1 solid H1',
#     #     'border-right': '2 dashed H2',
#     #     'border-left': '3 solid H2',
#     #     'margin': 1,
#     #     'padding': 2,
#     # },
#     {'border': '2 solid H1'},
#     {'padding': 2, 'ansi': 'H2'},
# )
# # normal text color:
# color = 'T'

# # terminal columns (width). 0 -> autodetect or use from cli
# col = 0

# # hirarchical indentation by:
# left_indent = '  '

# link_start = '①'
# link_start_ord = ord(link_start)

# log_level = 'debug'

# # it: inline table, h: hide, i: inline
# show_links = 'it'

# tab_length = 4

# # could be given, otherwise read from ansi_tables.json:
# themes = {}


# # sample for the theme roller feature:
# md_sample = ''

# # dir monitor recursion max:
# mon_max_files = 1000

# rm_spaces = True

# sample_text = '''
# # H1

# Hello *World!*
# '''

# ------------------------------------------------------------------ End Config
